File: podklady/6_Others/2_github_copilot_sdk/python/3_workflows/1_sequential_workflow.py
"""
Workflow — Sequential.

Concept
-------
Linear pipeline: research → analyse → summarise. Each stage's output is fed
into the next stage's prompt. No branching, no parallelism.

Run:
    python 3_workflows/1_sequential_workflow.py
"""

import asyncio

from copilot import CopilotClient
from copilot.session import PermissionHandler


async def run_stage(client: CopilotClient, role_prompt: str, user_prompt: str) -> str:
    session = await client.create_session(
        on_permission_request=PermissionHandler.approve_all,
        system_message={"mode": "replace", "content": role_prompt},
    )
    try:
        reply = await session.send_and_wait(user_prompt, timeout=300.0)
        return reply.data.content if reply else ""
    finally:
        await session.disconnect()


async def main() -> None:
    client = CopilotClient()
    await client.start()
    try:
        topic = "the practical benefits of asyncio for I/O-bound Python services"

        print("=== stage 1: research ===")
        findings = await run_stage(
            client,
            "You are a research analyst. Produce a dense bullet list of relevant facts.",
            f"Research: {topic}",
        )
        print(findings)

        print("\n=== stage 2: analysis ===")
        analysis = await run_stage(
            client,
            "You are a technical analyst. Identify trade-offs, risks, and key insights.",
            f"Analyse these findings:\n\n{findings}",
        )
        print(analysis)

        print("\n=== stage 3: summary ===")
        summary = await run_stage(
            client,
            "You are an executive writer. Produce a 5-line summary aimed at engineering managers.",
            f"Summarise this analysis:\n\n{analysis}",
        )
        print(summary)
    finally:
        await client.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: podklady/6_Others/2_github_copilot_sdk/typescript/src/3_workflows/1_sequential_workflow.ts
/**
 * Workflow — Sequential.
 *
 * Concept
 * -------
 * Linear pipeline: research → analyse → summarise. Each stage's output is
 * fed into the next stage's prompt. No branching, no parallelism.
 *
 * Run:
 *   npx tsx src/3_workflows/1_sequential_workflow.ts
 */

import { CopilotClient, approveAll } from "@github/copilot-sdk";

async function runStage(
    client: CopilotClient,
    rolePrompt: string,
    userPrompt: string,
): Promise<string> {
    const session = await client.createSession({
        onPermissionRequest: approveAll,
        systemMessage: { mode: "replace", content: rolePrompt },
    });
    try {
        const reply = await session.sendAndWait({ prompt: userPrompt }, 300_000);
        return (reply?.data as { content: string })?.content ?? "";
    } finally {
        await session.disconnect();
    }
}

async function main(): Promise<void> {
    const client = new CopilotClient();
    try {
        const topic = "the practical benefits of async/await for I/O-bound Node services";

        console.log("=== stage 1: research ===");
        const findings = await runStage(
            client,
            "You are a research analyst. Produce a dense bullet list of relevant facts.",
            `Research: ${topic}`,
        );
        console.log(findings);

        console.log("\n=== stage 2: analysis ===");
        const analysis = await runStage(
            client,
            "You are a technical analyst. Identify trade-offs, risks, and key insights.",
            `Analyse these findings:\n\n${findings}`,
        );
        console.log(analysis);

        console.log("\n=== stage 3: summary ===");
        const summary = await runStage(
            client,
            "You are an executive writer. Produce a 5-line summary aimed at engineering managers.",
            `Summarise this analysis:\n\n${analysis}`,
        );
        console.log(summary);
    } finally {
        await client.stop();
    }
}

main().catch(console.error);
